contracts/Staking.sol

```diff
@@ -0,0 +1,180 @@
+        SEURO_ADDRESS = _SEURO_ADDRESS;
+        startTime = _start;
+        endTime = _end;
+        initialised = block.timestamp;
+
+        // TODO variable
+        minTST = 1 ether;
+
+        active = true;
+    }
+
+    function disable() external onlyOwner {
+        require(active, 'err-not-active');
+        active = false;
+    }
+
+    // reward works out the amount of seuro given to the user based on the
+    // amount of TST they first put in.
+    function reward(uint256 _amount) public view returns (uint256) {
+        uint256 SEURO = _amount * SEUROTST / 10_000;
+        uint256 REWARD = SEURO * INTEREST / 10_000;
+        return REWARD + SEURO;
+    }
+
+    function mint(uint256 _amount) external returns(uint256) {
+
+        // TODO CHORE needs refactor
+        require(active == true, 'err-not-active');
+        require(_amount >= minTST, 'err-not-min');
+        require(block.timestamp >= startTime, 'err-not-started');
+        require(block.timestamp < endTime, 'err-finished');
+
+        // calculate the reward so we can also update the remaining SEURO
+        uint256 total = reward(_amount);
+        require(SEURO_REMAINING >= total, 'err-overlimit');
+
+        // Transfer funds from sender to this contract
+        // TODO send to some other guy not this contract!
+
+        IERC20 TOKEN = IERC20(TST_ADDRESS);
+        TOKEN.transferFrom(msg.sender, address(this), _amount);
+
+        // fetch current tokenID
+        uint256 newItemId = _tokenIds.current();
+
+        Position memory position = _positions[msg.sender];
+
+        if (position.nonce > 0) {
+            position.totalValue += _amount;
+            position.nonce += 1;
+            position.reward += total;
+        }
+
+        if (position.nonce == 0) {
+            _mint(msg.sender, newItemId);
+
+            position.nonce = 1;
+            position.open = true;
+            position.totalValue = _amount;
+            position.tokenId = newItemId;
+            position.reward = total;
+
+            _tokenIds.increment();
+        }
+
+        // update the position
+        _positions[msg.sender] = position;
+
+        // update the remaining SEURO
+        SEURO_REMAINING -= total;
+
+        // makes no sence to return this.
+        return newItemId;
+    }
+
+    function burn() public {
+        require(block.timestamp >= endTime, 'err-pool-open');
+
+        Position memory position = _positions[msg.sender];
+        require(position.nonce > 0, 'err-not-valid');
+        require(position.open == true, 'err-closed');
+
+        // update position
+        position.open = false;
+
+        // burn the token
+        _burn(position.tokenId);
+
+        // withdraw funds
+        IERC20 TOKEN = IERC20(SEURO_ADDRESS);
+        TOKEN.transfer(msg.sender, position.reward);
+        
+        _positions[msg.sender] = position;
+    }
+
+    // withdraw to the owners address
+    function withdraw(address _address) external onlyOwner {
```

Comment:
No, because this contract ONLY accepts TST. However, we need to top it up with SEURO.

So, we would need to extract TST after it's done. 

And we *might* need to extract the SEURO if there's an issue or we cancel it.

Reply:
what about making two withdraw functions: one for TST and one for SEURO? since you've already hardcoded their addresses in the constructor. then there won't be any ambiguity on as to what is being withdrawed. since this function only accepts two different types of address too.

Reply:
clarifications done in voice convo. it's useful to be able to withdraw any tokens that are sent to the contract, _despite_ only expecting TST and sEURO.